zilliqa/src/consensus.rs

```diff
@@ -419,11 +426,37 @@ impl Consensus {
             return Ok(None);
         }
 
+        let head_block = self.head_block();
+        let head_block_view = head_block.view();
+
+        if head_block_view + 1 == self.view.get_view() && self.create_next_block_on_timeout {
```

Comment:
I wonder if we can make the logic slightly more stateless here instead? I am worried keeping track of it in a separate flag might lead to weird bugs, especially when taking into accounts forks, new views, etc.

In what situations would we call `timeout()` and `create_next_block_on_timeout` be `false`?

Reply:
`empty_block_timeout` does not only have to be smaller than `consensus_timeout` , we also need to carefully determine the latest point when we would reach supermajority with an empty mempool and decide to wait for `empty_block_timeout` instead of proposing a block. This is what I tried to convey in my second comment below.